crawler/simple_crawler.py

```diff
@@ -0,0 +1,111 @@
+import hashlib
+import logging
+import os
+from urllib.parse import urlparse, urljoin
+
+from bs4 import BeautifulSoup
+
+logger = logging.getLogger('SimpleCrawler')
+
+
+class Response:
+    def __init__(self, status_code, reason):
+        self.text = None
+        self.status_code = status_code
+        self.reason = reason
+
+
+def open_file(url):
+    # path = os.path.normcase(os.path.normpath(url))
+
+    status_code, reason = check_path(url)
+    response = Response(status_code, reason)
+    if response.status_code == 200:
+        with open(url, 'rb') as f:
+            response.text = f.read()
+    return response
+
+
+def check_path(path):
+    """Return an HTTP status for the given filesystem path."""
+    if os.path.isdir(path):
+        return 400, "Path Not A File"
+    elif not os.path.isfile(path):
+        return 404, "File Not Found"
+    elif not os.access(path, os.R_OK):
+        return 403, "Access Denied"
+    else:
+        return 200, "OK"
+
+
+class SimpleCrawler:
+    def __init__(self, url):
+        self.start_page = url
+        self.visited_child_resources = {}
+        self.unreachable_child_resources = {}
+
+    class Resources:
+        def __init__(self, url, parent_url):
+            m = hashlib.sha256()
+            m.update(bytes(url, 'utf-8'))
+            self.id = m.hexdigest()
+            self.url = url
+            self.parent_url = parent_url
+
```

Comment:
these two methods are poorly named, in my opinion @rnllv 

They are called `add`,  but they also return true/false information.

I _think_ I'd rather not have the methods.

I _think_ it would be neater to have at line 82:
```
  child_resource_already_accessed = resource.id in self.visited_child_resources
  if not child_resource_already_accessed:   
     self.visited.child_resources[resource.id] = resource  
```

Reply:
>This is hidden behaviour.

Coming to think of it, yes, it is indeed hidden behaviour.

I'll incorporate the changes you've suggested.

Reply:
Ohh, I forgot. The method was intentional. I agree the name should be improved.

I need those methods there so that if we decide to take a multi-threaded approach in future (for performance), the body in these methods will become `critical sections` (since the dictionaries are shared)  and can be protected by a `lock` object to prevent race conditions.

Reply:
Isolating the `critical sections` in dedicated methods would be more easy to debug that to put it throughout the crawler.

Reply:
:)
Okay. Let me implement the code you suggested.

Reply:
Suggested review changes pushed. Please check.